parachain/runtime/rococo/src/weights/incentivized_channel_outbound_weights.rs

```diff
@@ -0,0 +1,53 @@
+
+//! Autogenerated weights for incentivized_channel::outbound
+//!
+//! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 3.0.0
+//! DATE: 2021-04-14, STEPS: `[50, ]`, REPEAT: 20, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! EXECUTION: Some(Wasm), WASM-EXECUTION: Compiled, CHAIN: Some("/tmp/artemis-benchmark-vAY/spec.json"), DB CACHE: 128
+
+// Executed Command:
+// target/release/artemis
+// benchmark
+// --chain
+// /tmp/artemis-benchmark-vAY/spec.json
+// --execution
+// wasm
+// --wasm-execution
+// compiled
+// --pallet
+// incentivized_channel::outbound
+// --extrinsic
+// *
+// --repeat
+// 20
+// --steps
+// 50
+// --output
+// runtime/rococo/src/weights/incentivized_channel_outbound_weights.rs
+
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::{traits::Get, weights::Weight};
+use sp_std::marker::PhantomData;
+
+/// Weight functions for incentivized_channel::outbound.
+pub struct WeightInfo<T>(PhantomData<T>);
+impl<T: frame_system::Config> incentivized_channel::outbound::WeightInfo for WeightInfo<T> {
+	fn on_initialize(m: u32, ) -> Weight {
```

Comment:
Just curious, how does the benchmarking weight generator know what to do with `m`/`num_messages` (i.e the saturating_mul)? I don't see some kind of template being used.

Reply:
This line in the benchmark is special macro syntax:
```
let m in 1 .. T::MaxMessagesPerCommit::get() as u32;
```

It declares `m` as a variable for the purpose of linear regression. This results in the benchmark being run for various values of `m` (in our case it will be run 21 times for each value of `m` due to the benchmark args `--steps 50 --repeat 20`). And then the linear regression analysis automatically figures out the multiplier for `m`.

Major caveat here: this means `frame-benchmarking` can only be used to benchmark O(1) and O(n) complexity. Logarithmic / superlinear complexity is off the table.

Reply:
awesome, very cool :)